### PR TITLE
Fix to call cDU when an instance calls setState

### LINKED ISSUE #1452: componentDidUpdate is not called for shallow rendered component
### npm package versions
enzyme: 3.2.0
enzyme-adapter-react-16: 1.1.1
react: 16.2.0
react-dom: 16.2.0

### Problem
When changing the state inside a component, `componentDidUpdate` is not called.

### Expected behaviour
When changing the state inside a component, `componentDidUpdate` should get called.

### Code example
```
import { shallow } from 'enzyme';
import React from 'react';

class Dropdown extends React.Component {
    constructor(props) {
        super(props);

        console.log('constructor');

        this.state = {
            highlightedIndex: 0,
        };

        this.handleMouseOver = this.handleMouseOver.bind(this);
    }

    componentDidUpdate() {
        console.log('componentDidUpdate');
    }

    handleMouseOver() {
        console.log('handleMouseOver', 0);

        this.setState({
            highlightedIndex: 1,
        }, () => {
            console.log('handleMouseOver', 1);
        });

        console.log('handleMouseOver', 2);
    }

    render() {
        console.log('render');

        return <div onMouseOver={this.handleMouseOver}>Please hover me</div>;
    }
}

it('should work', () => {
    const wrapper = shallow(<Dropdown />);

    wrapper.simulate('mouseover');
    wrapper.update();
});
```

#### Console output
```
  console.log .../__tests__/Dropdown.js:8
    constructor

  console.log .../__tests__/Dropdown.js:34
    render

  console.log .../__tests__/Dropdown.js:22
    handleMouseOver 0

  console.log .../__tests__/Dropdown.js:34
    render

  console.log .../__tests__/Dropdown.js:27
    handleMouseOver 1

  console.log .../__tests__/Dropdown.js:30
    handleMouseOver 2
```

### PR DESCRIPTION
**Pull request by koba04**
Fix #1452

This is not an ideal fix, but works 😅 

(This PR is based on #1741 to run all tests)

### CODE REVIEW

**Comment by koba04** [comment]
hmm, I should pass the test case 

https://github.com/airbnb/enzyme/pull/1742/commits/3b3ab38b83b59bdb55ea451331fcf0ced28f9abf

**Comment by koba04** [comment]
Fixed https://github.com/airbnb/enzyme/pull/1742#issuecomment-411633446

File: packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx [hunk 1]
```diff
@@ -4763,6 +4763,77 @@ describe('shallow', () => {
...
+        const wrapper = shallow(<Foo />);
+        wrapper.setState({ foo: 'wrapper setState update' });
+        expect(wrapper.state('foo')).to.equal('wrapper setState update');
+        expect(spy).to.have.property('callCount', 1);
+        wrapper.instance().onChange();
+        expect(wrapper.state('foo')).to.equal('onChange update');
+        expect(spy).to.have.property('callCount', 2);
+      });
+
+      it('should call `componentDidUpdate` when component’s `setState` is called through a bound method', () => {
```
[the remark this answers is not in the document]
Comment (koba04) [reply packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: This test is to verify to work well with a bound function.
`this.onChange = this.onChange.bind(this);`

My first implementation didn't pass this.

**Comment by ajbogh** [comment]
LGTM!

**Comment by ajbogh** [comment]
@koba04, I updated to v3.4.1 which should have included this change, but the issue still isn't fixed. Please see my example below:

```
// component code

  componentDidUpdate(prevProps: ViewProps, prevState: ViewState) {
    // this does not fire
    console.log("-----componentDidUpdate", prevProps.name, prevState.name, this.props.name, this.state.name);
  }

  changeName = (name: string) => {
    console.log("-------changeName"); // this fires
    this.setState({name});
  };


// Test code:

  fit('Calls setState if a name changed', () => {
    wrapper.instance().changeName('Something different');
    // even tried with wrapper.update() here
    expect(defaultProps.setUnsavedChanges).toHaveBeenCalledWith(true);
  });
```

**Comment by ajbogh** [comment]
Sure.

Please see the link below and feel free to clone the test repo included in the link to run the code.

https://github.com/airbnb/enzyme/issues/1755

File: packages/enzyme/src/ShallowWrapper.js [hunk 2]
```diff
@@ -170,6 +171,13 @@ class ShallowWrapper {
       privateSet(this, RENDERER, renderer);
       this[RENDERER].render(nodes, options.context);
       const { instance } = this[RENDERER].getNode();
+      const adapter = getAdapter(this[OPTIONS]);
+      const lifecycles = getAdapterLifecycles(adapter);
+      // Ensure to call componentDidUpdate when instance.setState is called
+      if (instance && lifecycles.componentDidUpdate.onSetState && !instance[SET_STATE]) {
+        privateSet(instance, SET_STATE, instance.setState);
+        instance.setState = (...args) => this.setState(...args);
```
Comment (mmairs9) [inline packages/enzyme/src/ShallowWrapper.js]: With this change i am now seeing alot of failing tests which have  a setState in componentDidMount. `Method "setState" is only meant to be run on a single node. undefined found instead.`

Comment (ljharb) [reply packages/enzyme/src/ShallowWrapper.js]: See #1756.

**Pull request merged**